aerogear-android-push/src/main/java/org/jboss/aerogear/android/unifiedpush/gcm/UnifiedPushInstanceIDListenerService.java

```diff
@@ -0,0 +1,138 @@
+/**
+ * JBoss, Home of Professional Open Source
+ * Copyright Red Hat, Inc., and individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * 	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.aerogear.android.unifiedpush.gcm;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.util.Base64;
+import android.util.Log;
+import com.google.android.gms.gcm.GoogleCloudMessaging;
+import com.google.android.gms.iid.InstanceID;
+import com.google.android.gms.iid.InstanceIDListenerService;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Map;
+import org.jboss.aerogear.android.core.Provider;
+import org.jboss.aerogear.android.pipe.http.HttpProvider;
+import org.jboss.aerogear.android.pipe.http.HttpRestProvider;
+import static org.jboss.aerogear.android.unifiedpush.gcm.AeroGearGCMPushRegistrar.REGISTRAR_PREFERENCE_PATTERN;
+
+/**
+ * This is an Android Service which listens for InstanceID messages from
+ * Google's GCM servicers.
+ *
+ * See
+ * https://developers.google.com/instance-id/guides/android-implementation#refresh_tokens
```

Comment:
It is a message from Google but it is undefined exactly what triggers it.  I read somewhere it can be triggered by backing up the application to Google's cloud then restoring your device and restoring the application.

I am not going to try and test that route because I saw a dragon along the path ;)